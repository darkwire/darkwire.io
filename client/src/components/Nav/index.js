import React, { Component } from 'react'
import PropTypes from 'prop-types'
import shortId from 'shortid'
import { Info, Settings, PlusCircle, User, Users, Lock, Unlock, Star } from 'react-feather'
import logoImg from 'img/logo.png'
import Dropdown, { DropdownTrigger, DropdownContent } from 'react-simple-dropdown'
import Username from 'components/Username'
import Clipboard from 'clipboard'
import $ from 'jquery';

class Nav extends Component {
  componentDidMount() {
    const clip = new Clipboard('.clipboard-trigger')

    clip.on('success', () => {
      $('.room-id').tooltip('show')
      setTimeout(() => {
        $('.room-id').tooltip('hide')
      }, 3000)
    })

    $(() => {
      $('.room-id').tooltip({
        trigger: 'manual',
      })
      $('.lock-room').tooltip({
        trigger: 'manual',
      })
    })
  }

  componentDidUpdate() {
    $(() => {
      $('.me-icon-wrap').tooltip()
      $('.owner-icon-wrap').tooltip()
    })
  }

  newRoom() {
    $('.navbar-collapse').collapse('hide')
    window.open(`/${shortId.generate()}`)
  }

  handleSettingsClick() {
    $('.navbar-collapse').collapse('hide')
    this.props.openModal('Settings')
  }

  handleAboutClick() {
    $('.navbar-collapse').collapse('hide')
    this.props.openModal('About')
  }

  handleToggleLock() {
    if (!this.props.iAmOwner) {
      $('.lock-room').tooltip('show')
      setTimeout(() => $('.lock-room').tooltip('hide'), 3000)
      return
    }
    this.props.toggleLockRoom()
  }

  render() {
    return (
      <nav className="navbar navbar-expand-md navbar-dark">
        <div className="meta">
          <img src={logoImg} alt="Darkwire" className="logo" />

          <button
            data-toggle="tooltip"
            data-placement="bottom"
            title={this.props.translations.copyButtonTooltip}
            data-clipboard-text={`${window.location.origin}/${this.props.roomId}`}
            className="btn btn-plain btn-link clipboard-trigger room-id ellipsis">
            {`/${this.props.roomId}`}
          </button>

          <span className="lock-room-container">
            <button
              onClick={this.handleToggleLock.bind(this)}
              className="lock-room btn btn-link btn-plain"
              data-toggle="tooltip"
              data-placement="bottom"
              title="You must be the owner to lock or unlock the room"
            >
              {this.props.roomLocked &&
                <Lock />
              }
              {!this.props.roomLocked &&
                <Unlock className="muted" />
              }
            </button>
          </span>

          <Dropdown className="members-dropdown">
            <DropdownTrigger>
              <button className="btn btn-link btn-plain members-action">
                <Users className="users-icon" />
              </button>
              <span>{this.props.members.length}</span>
            </DropdownTrigger>
            <DropdownContent>
              <ul className="plain">
                {this.props.members.map((member, index) => (
                  <li key={`user-${index}`}>
                    <Username username={member.username} />
                    <span className="icon-container">
                      {member.id === this.props.userId &&
                        <span data-toggle="tooltip" data-placement="bottom" title="Me" className="me-icon-wrap">
                          <User className="me-icon" />
                        </span>
                      }
                      {member.isOwner &&
                        <span data-toggle="tooltip" data-placement="bottom" title="Owner" className="owner-icon-wrap">
                          <Star className="owner-icon" />
                        </span>
                      }
                    </span>
                  </li>
                ))}
              </ul>
            </DropdownContent>
          </Dropdown>
        </div>

        <button
          className="navbar-toggler"
          type="button"
          data-toggle="collapse"
          data-target="#navbarSupportedContent"
          aria-controls="navbarSupportedContent"
          aria-expanded="false"
          aria-label="Toggle navigation">
          <span className="navbar-toggler-icon" />
        </button>
        <div className="collapse navbar-collapse" id="navbarSupportedContent">
          <ul className="navbar-nav ml-auto">
            <li className="nav-item">
              <button className="btn btn-plain nav-link" onClick={this.newRoom.bind(this)}target="blank"><PlusCircle /> <span>{this.props.translations.newRoomButton}</span></button>
            </li>
            <li className="
            nav-item">
              <button onClick={this.handleSettingsClick.bind(this)} className="btn btn-plain nav-link"><Settings /> <span>{this.props.translations.settingsButton}</span></button>
            </li>
            <li className="nav-item">
              <button onClick={this.handleAboutClick.bind(this)} className="btn btn-plain nav-link"><Info /> <span>{this.props.translations.aboutButton}</span></button>
            </li>
          </ul>
        </div>
      </nav>
    )
  }
}

Nav.propTypes = {
  members: PropTypes.array.isRequired,
  roomId: PropTypes.string.isRequired,
  userId: PropTypes.string.isRequired,
  roomLocked: PropTypes.bool.isRequired,
  toggleLockRoom: PropTypes.func.isRequired,
  openModal: PropTypes.func.isRequired,
  iAmOwner: PropTypes.bool.isRequired,
  translations: PropTypes.object.isRequired,
}

export default Nav
