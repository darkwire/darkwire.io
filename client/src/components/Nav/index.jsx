import React from 'react';
import PropTypes from 'prop-types';
import { nanoid } from 'nanoid';
import { Info, Settings, PlusCircle, User, Users, Lock, Unlock, Star } from 'react-feather';
import $ from 'jquery';
import { Tooltip } from 'react-tooltip';
import { useClickOutside, useSafeState } from '@react-hookz/web/esnext';

import logoImg from '@/img/logo.png';
import Username from '@/components/Username';

const Nav = ({ members, roomId, userId, roomLocked, toggleLockRoom, openModal, iAmOwner, translations }) => {
  const [showCopyTooltip, setShowCopyTooltip] = useSafeState(false);
  const [showLockedTooltip, setShowLockedTooltip] = useSafeState(false);
  const [showMemberList, setShowMemberList] = useSafeState(false);
  const userListRef = React.useRef(null);
  const roomUrl = `${window.location.origin}/${roomId}`;

  useClickOutside(userListRef, () => {
    setShowMemberList(false);
  });

  const newRoom = () => {
    $('.navbar-collapse').collapse('hide');
    window.open(`/${nanoid()}`);
  };

  const handleSettingsClick = () => {
    $('.navbar-collapse').collapse('hide');
    openModal('Settings');
  };

  const handleAboutClick = () => {
    $('.navbar-collapse').collapse('hide');
    openModal('About');
  };

  const handleToggleLock = () => {
    if (!iAmOwner) {
      setShowLockedTooltip(true);
      setTimeout(() => {
        setShowLockedTooltip(false);
      }, 2000);
    } else {
      toggleLockRoom();
    }
  };

  const handleCopy = async () => {
    await navigator.clipboard.writeText(roomUrl);
    setShowCopyTooltip(true);
    setTimeout(() => {
      setShowCopyTooltip(false);
    }, 2000);
  };

  return (
    <nav className="navbar navbar-expand-md navbar-dark">
      <div className="meta">
        <img src={logoImg} alt="Darkwire" className="logo" />

        <button
          id="copy-room-url-button"
          className="btn btn-plain btn-link clipboard-trigger room-id ellipsis"
          onClick={handleCopy}
        >
          {`/${roomId}`}
        </button>
        {showCopyTooltip && (
          <Tooltip
            anchorId="copy-room-url-button"
            content={translations.copyButtonTooltip}
            place="bottom"
            events={[]}
            isOpen={true}
          />
        )}
        <span className="lock-room-container">
          <button
            id="lock-room-button"
            data-testid="lock-room-button"
            onClick={handleToggleLock}
            className="lock-room btn btn-link btn-plain"
          >
            {roomLocked && <Lock />}
            {!roomLocked && <Unlock className="muted" />}
          </button>
          {showLockedTooltip && (
            <Tooltip
              anchorId="lock-room-button"
              content="You must be the owner to lock or unlock the room"
              place="bottom"
              events={[]}
              isOpen={true}
            />
          )}
        </span>
        <div className="members-menu" ref={userListRef}>
          <button
            className="btn btn-link btn-plain members-action"
            title="Users"
            onClick={() => setShowMemberList(prev => !prev)}
          >
            <Users className="users-icon" />
          </button>
          <span>{members.length}</span>

          {showMemberList && (
            <ul className="member-list">
              {members.map((member, index) => (
                <li key={`user-${index}`}>
                  <Username username={member.username} />
                  <span className="icon-container">
                    {member.id === userId && (
                      <span className="me-icon-wrap" title="Me">
                        <User className="me-icon" />
                      </span>
                    )}
                    {member.isOwner && (
                      <span className="owner-icon-wrap">
                        <Star className="owner-icon" title="Owner" />
                      </span>
                    )}
                  </span>
                </li>
              ))}
            </ul>
          )}
        </div>
      </div>

      <button
        className="navbar-toggler"
        type="button"
        data-toggle="collapse"
        data-target="#navbarSupportedContent"
        aria-controls="navbarSupportedContent"
        aria-expanded="false"
        aria-label="Toggle navigation"
      >
        <span className="navbar-toggler-icon" />
      </button>
      <div className="collapse navbar-collapse" id="navbarSupportedContent">
        <ul className="navbar-nav ml-auto">
          <li className="nav-item">
            <button className="btn btn-plain nav-link" onClick={newRoom} target="blank">
              <PlusCircle /> <span>{translations.newRoomButton}</span>
            </button>
          </li>
          <li className="nav-item">
            <button onClick={handleSettingsClick} className="btn btn-plain nav-link">
              <Settings /> <span>{translations.settingsButton}</span>
            </button>
          </li>
          <li className="nav-item">
            <button onClick={handleAboutClick} className="btn btn-plain nav-link">
              <Info /> <span>{translations.aboutButton}</span>
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
};

Nav.propTypes = {
  members: PropTypes.array.isRequired,
  roomId: PropTypes.string.isRequired,
  userId: PropTypes.string.isRequired,
  roomLocked: PropTypes.bool.isRequired,
  toggleLockRoom: PropTypes.func.isRequired,
  openModal: PropTypes.func.isRequired,
  iAmOwner: PropTypes.bool.isRequired,
  translations: PropTypes.object.isRequired,
};

export default Nav;
