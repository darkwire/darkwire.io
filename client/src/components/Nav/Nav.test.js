import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import Nav from '.';
import mock$ from 'jquery';

const mockTooltip = jest.fn().mockImplementation(param => {
  // console.log('tooltip', param);
});

const mockCollapse = jest.fn().mockImplementation(param => {
  // console.log('collapse', param);
});

jest.mock('jquery', () => {
  return jest.fn().mockImplementation(param => {
    // console.log('$', param);
    if (typeof param === 'function') {
      param();
    }
    return {
      tooltip: mockTooltip,
      collapse: mockCollapse,
    };
  });
});

jest.mock('shortid', () => {
  return {
    generate() {
      return 'fakeid';
    },
  };
});

jest.useFakeTimers();

const mockTranslations = {
  newRoomButton: 'new room',
  settingsButton: 'settings',
  aboutButton: 'about',
};

test('Nav component is displaying', async () => {
  const { asFragment } = render(
    <Nav
      members={[]}
      roomId={'testRoom'}
      userId={'userId__'}
      roomLocked={false}
      toggleLockRoom={() => {}}
      openModal={() => {}}
      iAmOwner={true}
      translations={{}}
    />,
  );

  expect(asFragment()).toMatchSnapshot();

  expect(mock$).toHaveBeenCalledWith('.room-id');
  expect(mock$).toHaveBeenLastCalledWith('.lock-room');
  expect(mockTooltip).toHaveBeenLastCalledWith({ trigger: 'manual' });
});

test('Nav component is displaying with another configuration and can rerender', async () => {
  const { asFragment, rerender } = render(
    <Nav
      members={[
        { id: 'id1', username: 'alan', isOwner: true },
        { id: 'id2', username: 'dan', isOwner: false },
      ]}
      roomId={'testRoom_2'}
      userId={'userId_2'}
      roomLocked={true}
      toggleLockRoom={() => {}}
      openModal={() => {}}
      iAmOwner={false}
      translations={{}}
    />,
  );

  expect(asFragment()).toMatchSnapshot();

  rerender(
    <Nav
      members={[
        { id: 'id1', username: 'alan', isOwner: true },
        { id: 'id2', username: 'dan', isOwner: false },
      ]}
      roomId={'testRoom_3'}
      userId={'userId_3'}
      roomLocked={true}
      toggleLockRoom={() => {}}
      openModal={() => {}}
      iAmOwner={false}
      translations={{}}
    />,
  );

  expect(mock$).toHaveBeenCalledWith('.me-icon-wrap');
  expect(mock$).toHaveBeenLastCalledWith('.owner-icon-wrap');
});

test('Can copy room url', async () => {
  document.execCommand = jest.fn(() => true);

  const toggleLockRoom = jest.fn();

  const { getByText } = render(
    <Nav
      members={[
        { id: 'id1', username: 'alan', isOwner: true },
        { id: 'id2', username: 'dan', isOwner: false },
      ]}
      roomId={'testRoom'}
      userId={'userId'}
      roomLocked={true}
      toggleLockRoom={toggleLockRoom}
      openModal={() => {}}
      iAmOwner={false}
      translations={{}}
    />,
  );

  fireEvent.click(getByText(`/testRoom`));

  expect(document.execCommand).toHaveBeenLastCalledWith('copy');
  expect(mock$).toHaveBeenCalledTimes(15);
  expect(mockTooltip).toHaveBeenLastCalledWith('show');

  // Wait tooltip closing
  jest.runAllTimers();

  expect(mock$).toHaveBeenCalledTimes(18);
  expect(mockTooltip).toHaveBeenLastCalledWith('hide');
});

test('Can lock/unlock room is room owner only', async () => {
  const toggleLockRoom = jest.fn();

  const { rerender, getByTitle } = render(
    <Nav
      members={[
        { id: 'id1', username: 'alan', isOwner: true },
        { id: 'id2', username: 'dan', isOwner: false },
      ]}
      roomId={'testRoom'}
      userId={'userId'}
      roomLocked={true}
      toggleLockRoom={toggleLockRoom}
      openModal={() => {}}
      iAmOwner={true}
      translations={{}}
    />,
  );

  const toggleLockRoomButton = getByTitle('You must be the owner to lock or unlock the room');

  fireEvent.click(toggleLockRoomButton);

  expect(toggleLockRoom).toHaveBeenCalledWith();

  fireEvent.click(toggleLockRoomButton);

  expect(toggleLockRoom).toHaveBeenCalledTimes(2);

  // We are not the room owner anymore
  rerender(
    <Nav
      members={[
        { id: 'id1', username: 'alan', isOwner: true },
        { id: 'id2', username: 'dan', isOwner: false },
      ]}
      roomId={'testRoom'}
      userId={'userId'}
      roomLocked={true}
      toggleLockRoom={toggleLockRoom}
      openModal={() => {}}
      iAmOwner={false}
      translations={{}}
    />,
  );

  fireEvent.click(toggleLockRoomButton);

  expect(toggleLockRoom).toHaveBeenCalledTimes(2);
  expect(mock$).toHaveBeenLastCalledWith('.lock-room');
  expect(mockTooltip).toHaveBeenLastCalledWith('show');
});

test('Can show user list', async () => {
  // Test with one user owner and me
  const { getByTitle, getByText, queryByTitle, rerender } = render(
    <Nav
      members={[{ id: 'id1', username: 'alan', isOwner: true }]}
      roomId={'testRoom'}
      userId={'id1'}
      roomLocked={true}
      toggleLockRoom={() => {}}
      openModal={() => {}}
      iAmOwner={true}
      translations={{}}
    />,
  );

  fireEvent.click(getByTitle('Users'));

  await waitFor(() => expect(getByText('alan')).toBeInTheDocument());
  await waitFor(() => expect(getByTitle('Owner')).toBeInTheDocument());
  await waitFor(() => expect(getByTitle('Me')).toBeInTheDocument());

  // Test with two user not owner, not me
  rerender(
    <Nav
      members={[
        { id: 'id1', username: 'alan', isOwner: false },
        { id: 'id2', username: 'dan', isOwner: false },
      ]}
      roomId={'testRoom'}
      userId={'otherId'}
      roomLocked={true}
      toggleLockRoom={() => {}}
      openModal={() => {}}
      iAmOwner={true}
      translations={{}}
    />,
  );
  await waitFor(() => expect(getByText('alan')).toBeInTheDocument());
  await waitFor(() => expect(getByText('dan')).toBeInTheDocument());
  expect(queryByTitle('Owner')).not.toBeInTheDocument();
  expect(queryByTitle('Me')).not.toBeInTheDocument();
});

test('Can open settings', async () => {
  const openModal = jest.fn();

  // Test with one user owner and me
  const { getByText } = render(
    <Nav
      members={[]}
      roomId={'testRoom'}
      userId={'id1'}
      roomLocked={true}
      toggleLockRoom={() => {}}
      openModal={openModal}
      iAmOwner={true}
      translations={mockTranslations}
    />,
  );

  fireEvent.click(getByText(mockTranslations.settingsButton));

  expect(mock$).toHaveBeenLastCalledWith('.navbar-collapse');
  expect(mockCollapse).toHaveBeenLastCalledWith('hide');
  expect(openModal).toHaveBeenLastCalledWith('Settings');
});

test('Can open About', async () => {
  const openModal = jest.fn();

  // Test with one user owner and me
  const { getByText } = render(
    <Nav
      members={[]}
      roomId={'testRoom'}
      userId={'id1'}
      roomLocked={true}
      toggleLockRoom={() => {}}
      openModal={openModal}
      iAmOwner={true}
      translations={mockTranslations}
    />,
  );

  fireEvent.click(getByText(mockTranslations.aboutButton));

  expect(mock$).toHaveBeenLastCalledWith('.navbar-collapse');
  expect(mockCollapse).toHaveBeenLastCalledWith('hide');
  expect(openModal).toHaveBeenLastCalledWith('About');
});

test('Can open About', async () => {
  window.open = jest.fn();

  // Test with one user owner and me
  const { getByText } = render(
    <Nav
      members={[]}
      roomId={'testRoom'}
      userId={'id1'}
      roomLocked={true}
      toggleLockRoom={() => {}}
      openModal={() => {}}
      iAmOwner={true}
      translations={mockTranslations}
    />,
  );

  fireEvent.click(getByText(mockTranslations.newRoomButton));

  expect(mock$).toHaveBeenLastCalledWith('.navbar-collapse');
  expect(mockCollapse).toHaveBeenLastCalledWith('hide');
  expect(window.open).toHaveBeenLastCalledWith('/fakeid');
});
