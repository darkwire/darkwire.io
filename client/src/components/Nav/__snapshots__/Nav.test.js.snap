// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav component is displaying 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-md navbar-dark"
  >
    <div
      class="meta"
    >
      <img
        alt="Darkwire"
        class="logo"
        src="logo.png"
      />
      <button
        class="btn btn-plain btn-link clipboard-trigger room-id ellipsis"
        data-clipboard-text="http://localhost/testRoom"
        data-placement="bottom"
        data-toggle="tooltip"
      >
        /testRoom
      </button>
      <span
        class="lock-room-container"
      >
        <button
          class="lock-room btn btn-link btn-plain"
          data-placement="bottom"
          data-toggle="tooltip"
          title="You must be the owner to lock or unlock the room"
        >
          <svg
            class="muted"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="11"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="11"
            />
            <path
              d="M7 11V7a5 5 0 0 1 9.9-1"
            />
          </svg>
        </button>
      </span>
      <div
        class="dropdown members-dropdown"
      >
        <a
          class="dropdown__trigger "
        >
          <button
            class="btn btn-link btn-plain members-action"
            title="Users"
          >
            <svg
              class="users-icon"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 21v-2a4 4 0 0 0-4-4H5a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="9"
                cy="7"
                r="4"
              />
              <path
                d="M23 21v-2a4 4 0 0 0-3-3.87"
              />
              <path
                d="M16 3.13a4 4 0 0 1 0 7.75"
              />
            </svg>
          </button>
          <span>
            0
          </span>
        </a>
        <div
          class="dropdown__content "
        >
          <ul
            class="plain"
          />
        </div>
      </div>
    </div>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      class="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav ml-auto"
      >
        <li
          class="nav-item"
        >
          <button
            class="btn btn-plain nav-link"
            target="blank"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="16"
              />
              <line
                x1="8"
                x2="16"
                y1="12"
                y2="12"
              />
            </svg>
             
            <span />
          </button>
        </li>
        <li
          class=" nav-item"
        >
          <button
            class="btn btn-plain nav-link"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="3"
              />
              <path
                d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
              />
            </svg>
             
            <span />
          </button>
        </li>
        <li
          class="nav-item"
        >
          <button
            class="btn btn-plain nav-link"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="12"
                x2="12"
                y1="16"
                y2="12"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="8"
              />
            </svg>
             
            <span />
          </button>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Nav component is displaying with another configuration and can rerender 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-md navbar-dark"
  >
    <div
      class="meta"
    >
      <img
        alt="Darkwire"
        class="logo"
        src="logo.png"
      />
      <button
        class="btn btn-plain btn-link clipboard-trigger room-id ellipsis"
        data-clipboard-text="http://localhost/testRoom_2"
        data-placement="bottom"
        data-toggle="tooltip"
      >
        /testRoom_2
      </button>
      <span
        class="lock-room-container"
      >
        <button
          class="lock-room btn btn-link btn-plain"
          data-placement="bottom"
          data-toggle="tooltip"
          title="You must be the owner to lock or unlock the room"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="11"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="11"
            />
            <path
              d="M7 11V7a5 5 0 0 1 10 0v4"
            />
          </svg>
        </button>
      </span>
      <div
        class="dropdown members-dropdown"
      >
        <a
          class="dropdown__trigger "
        >
          <button
            class="btn btn-link btn-plain members-action"
            title="Users"
          >
            <svg
              class="users-icon"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 21v-2a4 4 0 0 0-4-4H5a4 4 0 0 0-4 4v2"
              />
              <circle
                cx="9"
                cy="7"
                r="4"
              />
              <path
                d="M23 21v-2a4 4 0 0 0-3-3.87"
              />
              <path
                d="M16 3.13a4 4 0 0 1 0 7.75"
              />
            </svg>
          </button>
          <span>
            2
          </span>
        </a>
        <div
          class="dropdown__content "
        >
          <ul
            class="plain"
          >
            <li>
              <span
                class="username"
                style="color: rgb(192, 202, 249);"
              >
                alan
              </span>
              <span
                class="icon-container"
              >
                <span
                  class="owner-icon-wrap"
                  data-placement="bottom"
                  data-toggle="tooltip"
                  title="Owner"
                >
                  <svg
                    class="owner-icon"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <polygon
                      points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
                    />
                  </svg>
                </span>
              </span>
            </li>
            <li>
              <span
                class="username"
                style="color: rgb(156, 252, 223);"
              >
                dan
              </span>
              <span
                class="icon-container"
              />
            </li>
          </ul>
        </div>
      </div>
    </div>
    <button
      aria-controls="navbarSupportedContent"
      aria-expanded="false"
      aria-label="Toggle navigation"
      class="navbar-toggler"
      data-target="#navbarSupportedContent"
      data-toggle="collapse"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
      id="navbarSupportedContent"
    >
      <ul
        class="navbar-nav ml-auto"
      >
        <li
          class="nav-item"
        >
          <button
            class="btn btn-plain nav-link"
            target="blank"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="16"
              />
              <line
                x1="8"
                x2="16"
                y1="12"
                y2="12"
              />
            </svg>
             
            <span />
          </button>
        </li>
        <li
          class=" nav-item"
        >
          <button
            class="btn btn-plain nav-link"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="3"
              />
              <path
                d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
              />
            </svg>
             
            <span />
          </button>
        </li>
        <li
          class="nav-item"
        >
          <button
            class="btn btn-plain nav-link"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="12"
                x2="12"
                y1="16"
                y2="12"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="8"
              />
            </svg>
             
            <span />
          </button>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;
